packages/core/src/directory.ts

```diff
@@ -0,0 +1,107 @@
+import { Org, OrgID, OrgMember, OrgMemberStatus } from "./org";
+import { Server } from "./server";
+
+export interface DirectoryUser {
+    externalId: string;
+    email: string;
+    name: string;
+    active: boolean;
+}
+
+export interface DirectoryGroup {
+    name: string;
+    members: DirectoryUser[];
+}
+
+export interface DirectorySubscriber {
+    userCreated(user: DirectoryUser, orgId: OrgID): Promise<void>;
+    userUpdated(user: DirectoryUser, orgId: OrgID): Promise<void>;
+    userDeleted(user: DirectoryUser, orgId: OrgID): Promise<void>;
+
+    groupCreated(group: DirectoryGroup, orgId: OrgID): Promise<void>;
+    groupUpdated(group: DirectoryGroup, orgId: OrgID): Promise<void>;
+    groupDeleted(group: DirectoryGroup, orgId: OrgID): Promise<void>;
+}
+
+export interface DirectoryProvider {
+    subscribe(sub: DirectorySubscriber): void;
+}
+
+export class DirectorySync implements DirectorySubscriber {
+    constructor(public readonly server: Server, providers: DirectoryProvider[] = []) {
+        for (const provider of providers) {
+            provider.subscribe(this);
+        }
+    }
+
+    async userCreated(user: DirectoryUser, orgId: string) {
+        const org = (orgId && (await this.server.storage.get(Org, orgId))) || null;
+        if (org && org.directory.syncProvider === "scim" && org.directory.syncMembers) {
+            const memberExists = org.members.some((member) => member.directoryExternalId === user.externalId);
+
+            if (memberExists) {
+                return;
+            }
```

Comment:
Without this we just kept creating the same user, which isn't a good idea since AD can send the same request more than once (if it failed or something).